core/model/licensing.py

```diff
@@ -2061,3 +2063,130 @@ def rights_uri_from_string(cls, rights):
+        :param prioritized_content_types: The set of content types to prioritize; items earlier in the list are higher priority.
+        :param hidden_content_types: The set of content types to remove entirely
+        """
+
+        # Assign priorities to each content type and DRM scheme based on their position
+        # in the given lists. Higher priorities are assigned to items that appear earlier.
+        self._prioritized_content_types = {}
+        _priority = 1
+        for content_type in reversed(prioritized_content_types):
+            self._prioritized_content_types[content_type] = _priority
+            _priority = _priority + 1
+
+        self._prioritized_drm_schemes = {}
+        _priority = 1
+        for drm_scheme in reversed(prioritized_drm_schemes):
+            self._prioritized_drm_schemes[drm_scheme] = _priority
+            _priority = _priority + 1
+
+        self._hidden_content_types = hidden_content_types
+
+    def prioritize_for_pool(
+        self, pool: LicensePool
+    ) -> List[LicensePoolDeliveryMechanism]:
+        """
+        Filter and prioritize the delivery mechanisms in the given pool.
+        :param pool: The license pool
+        :return: A list of suitable delivery mechanisms in priority order, highest priority first
+        """
+        return self.prioritize_mechanisms(pool.delivery_mechanisms)
+
+    def prioritize_mechanisms(
+        self, mechanisms: List[LicensePoolDeliveryMechanism]
+    ) -> List[LicensePoolDeliveryMechanism]:
+        """
+        Filter and prioritize the delivery mechanisms in the given pool.
+        :param mechanisms: The list of delivery mechanisms
+        :return: A list of suitable delivery mechanisms in priority order, highest priority first
+        """
+
+        # First, filter out all hidden content types.
+        mechanisms_filtered: List[LicensePoolDeliveryMechanism] = []
+        for delivery in mechanisms:
+            delivery_mechanism = delivery.delivery_mechanism
+            if delivery_mechanism:
+                if delivery_mechanism.content_type not in self._hidden_content_types:
+                    mechanisms_filtered.append(delivery)
+
+        # If there are any prioritized DRM schemes or content types, then
+        # sort the list of mechanisms accordingly.
+        if (
+            len(self._prioritized_drm_schemes) != 0
+            or len(self._prioritized_content_types) != 0
+        ):
+            mechanisms_filtered.sort(
+                key=cmp_to_key(self._compare_delivery_mechanisms), reverse=True
+            )
+
+        return mechanisms_filtered
+
+    def _drm_scheme_priority(self, drm_scheme: str) -> int:
+        """Determine the priority of a DRM scheme. A lack of DRM is always
+        prioritized over having DRM, and prioritized schemes are always
+        higher priority than non-prioritized schemes."""
+
+        if not drm_scheme:
+            return sys.maxsize
```

Comment:
The policy is very much open to changes. I think I mentioned on Slack; the choice to prioritize DRM-free content over DRM'd content if both types exist for a given book seemed more patron-friendly. Given the arbitrary nature of the input, it also seemed more useful to be able to say "DRM-free content will come first if it's there" rather than "It could be anywhere!".

Reply:
I agree that presenting DRM free content first is more patron friendly. What concerns me though, if is I have some content that has the following DRM types and are currently delivered in this order:
- `application/vnd.adobe.adept+xml`
- `application/vnd.readium.lcp.license.v1.0+json`
- DRM Free

And then I add a config setting to prefer `application/vnd.readium.lcp.license.v1.0+json`, but because we only prefer DRM free after there is a preference set, we will give the content in this order:
- DRM Free
- `application/vnd.readium.lcp.license.v1.0+json`
- `application/vnd.adobe.adept+xml`

This seems unexpected to me. You explicitly set a setting to prefer `application/vnd.readium.lcp.license.v1.0+json`, but that caused this title to prefer DRM free.

Reply:
I was trying to reason about this a bit, and its hard without knowing the default ordering. Did a bit of digging, and it seems like the [`initialize_data` function](https://github.com/ThePalaceProject/circulation/blob/ea393a6b4f95ea2a6fc9060db44759d2f6eb020e/core/model/__init__.py#L430-L439) is implicitly setting this order by creating the Delivery mechanisms in [`default_client_can_fulfill_lookup`](https://github.com/ThePalaceProject/circulation/blob/6c287c912b4992e5af1a424eca16d90beef884d1/core/model/licensing.py#L1756) when the database is first initialized. So that order is likely (maybe, probably) consistent in our existing DBs, and its prioritizing items without DRM as well. So my objection above can't actually happen.

Reply:
Yeah, the problem is that the default ordering is unpredictable. I can't promise the ordering you see in your database will be anything like the ordering I see in mine. It seems very much subject to what order content appeared in the original feeds, and the order in which you imported various collections.

Does anyone actually have books that appear as both DRM-free and DRM'd? That would seem to be ... kind of silly.

Reply:
(At least when they choose to opt-in by adding configuration values)

Reply:
I feel like maybe we could just eliminate any surprises with documentation.

Reply:
How do you feel about this as a fair warning?

![warn](https://user-images.githubusercontent.com/612494/153049631-d3005a78-4cb0-461b-9c75-d6390e34d248.png)